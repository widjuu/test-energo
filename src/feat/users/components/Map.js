import { MapContainer, TileLayer, Marker, Popup } from "react-leaflet";
import { useSelector } from "react-redux";

export const Map = () => {
  const selectedUsers = useSelector((state) => state.selectedUsers.selected);

  console.log(selectedUsers);

  const markers = Object.keys(selectedUsers).map((userId) => {
    const user = selectedUsers[userId];

    if (!user) {
      return null;
    }

    const lat = user?.address?.geo?.lat ?? 0;
    const lng = user?.address?.geo?.lng ?? 0;
    const name = user?.name ?? "Без имени";

    return (
      <Marker position={[lat, lng]} key={user.id}>
        <Popup>{name}</Popup>
      </Marker>
    );
  });

  console.log(markers);

  return (
    <MapContainer
      center={[0, 0]}
      zoom={1}
      scrollWheelZoom={false}
      className="leaflet"
    >
      <TileLayer
        attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      {markers}
    </MapContainer>
  );
};
