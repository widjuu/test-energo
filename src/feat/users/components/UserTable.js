import { useDispatch, useSelector } from "react-redux";

//components
import { Table } from "antd";

//utils
import { toggleUser } from "../actions";

const columns = [
  {
    title: "Имя",
    dataIndex: "name",
  },
  {
    title: "Телефон",
    dataIndex: "phone",
  },
];

export const UserTable = () => {
  const dispatch = useDispatch();
  const data = useSelector((state) => state.users.data);
  const selectedUsers = useSelector((state) => state.selectedUsers.selected);

  const dataSource = data.map((userId) => {
    return {
      id: userId?.id,
      name: userId?.name,
      key: userId?.id,
      phone: userId?.phone,
      address: userId?.address,
    };
  });

  const selectedRowKeys = Object.keys(selectedUsers).map((userId) => {
    const user = selectedUsers[userId];

    return user.id;
  });

  const rowSelection = {
    selectedRowKeys,
    onSelect(record) {
      dispatch(
        toggleUser({
          id: record.id,
          zipcode: record.zipcode,
          address: record.address,
        })
      );
    },
  };

  return (
    <div>
      <Table
        size="small"
        pagination={false}
        columns={columns}
        dataSource={dataSource}
        rowSelection={rowSelection}
      />
    </div>
  );
};
