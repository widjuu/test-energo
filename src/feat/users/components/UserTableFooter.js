import { useSelector, useDispatch } from "react-redux";

// components
import { makeStyles } from "@material-ui/core/styles";
import Box from "@material-ui/core/Box";
import Button from "@material-ui/core/Button";
import CircularProgress from "@material-ui/core/CircularProgress";

// redux
import { sendSelectedUsers } from "../actions";

const useStyles = makeStyles(({ palette }) => ({
  error: {
    color: palette.error.main,
  },
}));

export const UserTableFooter = () => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const selectedUsers = useSelector((state) => state.selectedUsers.selected);
  const loading = useSelector((state) => state.selectedUsers.loading);
  const error = useSelector((state) => state.selectedUsers.error);

  const disabled = Object.keys(selectedUsers).length === 0;

  const onButtonClick = () => dispatch(sendSelectedUsers(selectedUsers));

  return (
    <Box display="flex" alignItems="center" px={3} my={1}>
      <Button
        startIcon={
          loading ? <CircularProgress size={20} color="secondary" /> : undefined
        }
        onClick={onButtonClick}
        disabled={disabled}
        variant="contained"
        color={error ? "secondary" : "primary"}
      >
        Отправить
      </Button>
      {error ? (
        <Box ml={2} className={classes.error}>
          {error}
        </Box>
      ) : null}
    </Box>
  );
};
