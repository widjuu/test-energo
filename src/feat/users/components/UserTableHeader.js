import TableHead from "@material-ui/core/TableHead";
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";

export const UserTableHeader = () => {
  return (
    <TableHead>
      <TableRow>
        <TableCell>Имя</TableCell>
        <TableCell align="right">Телефон</TableCell>
        <TableCell padding="checkbox"></TableCell>
      </TableRow>
    </TableHead>
  );
};
