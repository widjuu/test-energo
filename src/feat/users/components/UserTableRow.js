import { useDispatch, useSelector } from "react-redux";

// components
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";
import Checkbox from "@material-ui/core/Checkbox";

import { toggleUser } from "../actions";

export const UserTableRow = ({ id, name, phone, address }) => {
  const dispatch = useDispatch();
  const selectedUsers = useSelector((state) => state.selectedUsers.selected);
  const isSelected = !!selectedUsers[id];

  const onClick = () => dispatch(toggleUser({ id, zipcode: address?.zipcode }));

  return (
    <TableRow hover>
      <TableCell>{name}</TableCell>
      <TableCell align="right">{phone}</TableCell>
      <TableCell padding="checkbox">
        <Checkbox checked={isSelected} onChange={onClick} />
      </TableCell>
    </TableRow>
  );
};
