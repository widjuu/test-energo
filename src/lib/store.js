import { configureStore } from "@reduxjs/toolkit";
import logger from "redux-logger";

import api from "./api";
import { usersReducer, selectedUsersReducer } from "../feat/users/reducer";

const isDev = process.env.NODE_ENV !== "production";

const store = configureStore({
  reducer: {
    users: usersReducer.reducer,
    selectedUsers: selectedUsersReducer.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      thunk: {
        extraArgument: { api },
      },
    }).concat(isDev ? [logger] : []),
  devTools: isDev,
});

export default store;
