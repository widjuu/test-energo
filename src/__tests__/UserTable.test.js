import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";

import { Provider } from "react-redux";
import store from "../lib/store";
import { UserTable } from "../feat/users/components/UserTable";

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("handles click correctly", () => {
  act(() => {
    render(
      <Provider store={store}>
        <UserTable />
      </Provider>,
      container
    );
  });
  //userEvent.click(container.getAllByRole("checkbox"));
  //expect(container.querySelector("input")).toBeChecked();
});
