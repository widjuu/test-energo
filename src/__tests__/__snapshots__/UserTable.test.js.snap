// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Table Snapshot +++capturing User Table Snapshot 1`] = `
<div>
  <div
    className="ant-table-wrapper"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-small ant-table-empty"
        >
          <div
            className="ant-table-container"
          >
            <div
              className="ant-table-content"
              onScroll={[Function]}
              style={Object {}}
            >
              <table
                style={
                  Object {
                    "tableLayout": "auto",
                  }
                }
              >
                <colgroup>
                  <col
                    className="ant-table-selection-col"
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                </colgroup>
                <thead
                  className="ant-table-thead"
                >
                  <tr>
                    <th
                      className="ant-table-cell ant-table-selection-column"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <div
                        className="ant-table-selection"
                      >
                        <label
                          className="ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
                        >
                          <span
                            className="ant-checkbox ant-checkbox-disabled"
                            style={Object {}}
                          >
                            <input
                              checked={false}
                              className="ant-checkbox-input"
                              disabled={true}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyPress={[Function]}
                              onKeyUp={[Function]}
                              type="checkbox"
                            />
                            <span
                              className="ant-checkbox-inner"
                            />
                          </span>
                        </label>
                      </div>
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      Имя
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      Телефон
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-placeholder"
                    style={
                      Object {
                        "display": null,
                      }
                    }
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={3}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <div
                        className="ant-empty ant-empty-normal"
                      >
                        <div
                          className="ant-empty-image"
                        >
                          <svg
                            className="ant-empty-img-simple"
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              fill="none"
                              fillRule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                className="ant-empty-img-simple-ellipse"
                                cx="32"
                                cy="33"
                                rx="32"
                                ry="7"
                              />
                              <g
                                className="ant-empty-img-simple-g"
                                fillRule="nonzero"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  className="ant-empty-img-simple-path"
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          className="ant-empty-description"
                        >
                          No Data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
